pylops/LinearOperator.py

```diff
@@ -47,32 +48,211 @@ class LinearOperator(spLinearOperator):
         Operator contains a matrix that can be solved explicitly
-        (``True``) or not (``False``)
+        (``True``) or not (``False``). Defaults to ``False``.
     clinear : :obj:`bool`, optional
         .. versionadded:: 1.17.0
 
-        Operator is complex-linear.
+        Operator is complex-linear. Defaults to ``True``.
     name : :obj:`str`, optional
         .. versionadded:: 2.0.0
 
         Name of operator (to be used by :func:`pylops.utils.describe.describe`)
 
     """
 
-    def __init__(self, Op=None, explicit=False, clinear=None, name=None):
-        self.explicit = explicit
+    def __init__(
+        self,
+        Op=None,
+        dtype=None,
+        shape=None,
+        dims=None,
+        dimsd=None,
+        clinear=None,
+        explicit=None,
+        name=None,
+    ):
         if Op is not None:
             self.Op = Op
-            self.shape = self.Op.shape
-            self.dtype = self.Op.dtype
-            self.clinear = getattr(self.Op, "clinear", True)
+            # All Operators must have shape and dtype
+            dtype = self.Op.dtype if dtype is None else dtype
+            shape = self.Op.shape if shape is None else shape
+            # Optionally, some operators have other attributes
+            dims = getattr(Op, "dims", (Op.shape[1],)) if dims is None else dims
+            dimsd = getattr(Op, "dimsd", (Op.shape[0],)) if dimsd is None else dimsd
+            clinear = getattr(Op, "clinear", True) if clinear is None else clinear
+            explicit = (
+                getattr(self.Op, "explicit", False) if explicit is None else explicit
+            )
+            if explicit and hasattr(Op, "A"):
+                self.A = Op.A
+            name = getattr(Op, "name", None) if name is None else name
+
+        if dtype is not None:
+            self.dtype = dtype
+        if shape is not None:
+            self.shape = shape
+        if dims is not None:
+            self.dims = dims
+        if dimsd is not None:
+            self.dimsd = dimsd
         if clinear is not None:
             self.clinear = clinear
+        if explicit is not None:
+            self.explicit = explicit
         self.name = name
 
+    @property
+    def shape(self):
+        _shape = getattr(self, "_shape", None)
+        if _shape is None:  # Cannot find shape, falling back on dims and dimsd
+            dims = getattr(self, "_dims", None)
+            dimsd = getattr(self, "_dimsd", None)
+            if dims is None or dimsd is None:  # Cannot find both dims and dimsd, error
+                raise AttributeError(
+                    (
+                        f"'{self.__class__.__name__}' object has no attribute 'shape' "
+                        "nor both fallback attributes ('dims', 'dimsd')"
+                    )
+                )
+            _shape = (np.prod(dimsd), np.prod(dims))
+            self._shape = _shape  # Update to not redo everything above on next call
```

Comment:
why is this needed here but not outside of the if clause?

Reply:
A simple getter with `default_value` could look like this:
```python
@property
def shape():
    if getattr(self, "_shape", False):
        return default_value
    return self._shape
```
In the actual getter, all I something similar: I check that `self._shape` has not been initialized yet (for example, the `__init__` function in a new operator does not provide shape, but provides `dims`/`dimsd`), and if it hasn't, then it can try to fall back on `dims`/`dimsd`. If those values do exist, we can just store them the first time they are computer. Moreover, they are "legal" (a 2-tuple of ints which agree with `dims`/`dimsd`), so they can bypass the checks. It would be perfectly OK to do `self.shape = _shape` in line 131, but it would just execute more code for no added benefit.